File: lib/sanitize.js
/**
 * PII sanitization utilities.
 * Masks phone numbers, limits content previews, redacts sensitive fields.
 */

const PHONE_PATTERN = /(?:\+?1[\s.-]?)?(?:\(?\d{3}\)?[\s.-]?)\d{3}[\s.-]?\d{4}/g;

/**
 * Mask a phone number: "5551234567" -> "***4567"
 */
export function maskPhone(phone) {
  if (!phone) return '[no-phone]';
  const digits = phone.replace(/\D/g, '');
  if (digits.length < 4) return '****';
  return '***' + digits.slice(-4);
}

/**
 * Truncate content for safe logging: "long string..." -> "long str..."
 */
export function truncate(str, maxLen = 30) {
  if (!str) return '';
  const value = String(str);
  if (value.length <= maxLen) return value;
  return value.substring(0, maxLen) + '...';
}

/**
 * Mask a senior name for logs: "David Zuluaga" -> "David Z."
 */
export function maskName(name) {
  if (!name) return '[unknown]';
  const parts = name.split(' ');
  if (parts.length === 1) return parts[0];
  return parts[0] + ' ' + parts.slice(1).map(p => p[0] + '.').join(' ');
}

/**
 * Mask phone-looking substrings inside free-form error messages.
 */
export function maskPhonesInText(text) {
  if (text == null) return text;
  return String(text).replace(PHONE_PATTERN, match => maskPhone(match));
}

/**
 * Return a safe, non-PHI error summary for structured logs.
 */
export function sanitizeError(error) {
  if (!error) return null;
  return {
    name: error.name || 'Error',
    code: error.code,
    message: truncate(maskPhonesInText(error.message || String(error)), 160),
  };
}

File: pipecat/lib/sanitize.py
"""PII sanitization utilities.

Port of lib/sanitize.js — masks phone numbers, names, and truncates content
for safe logging.
"""

from __future__ import annotations

import re


def mask_phone(phone: str | None) -> str:
    """Mask a phone number: '+15551234567' → '***4567'."""
    if not phone:
        return "[no-phone]"
    digits = re.sub(r"\D", "", phone)
    if len(digits) < 4:
        return "****"
    return "***" + digits[-4:]


def mask_name(name: str | None) -> str:
    """Mask a name for logs: 'David Zuluaga' → 'David Z.'."""
    if not name:
        return "[unknown]"
    parts = name.split()
    if len(parts) == 1:
        return parts[0]
    return parts[0] + " " + " ".join(p[0] + "." for p in parts[1:])


def truncate(text: str | None, max_len: int = 30) -> str:
    """Truncate content for safe logging."""
    if not text:
        return ""
    if len(text) <= max_len:
        return text
    return text[:max_len] + "..."
